flocker/node/agents/functional/test_cinder.py

```diff
@@ -415,3 +419,117 @@ def test_disk_attachment_fails_with_conflicting_disk(self):
                 transient_states=(u'attaching',),
             )
         self.assertEqual(e.exception.unexpected_state, u'available')
+
+class CinderAttachmentFLOC2991Tests(SynchronousTestCase):
+    """
+    Cinder volumes can be attached and return correct device path.
+    """
+    def setUp(self):
+        clients = openstack_clients()
+        self.cinder = clients['cinder_client']
+        self.nova = clients['nova_client']
+        self.blockdevice_api = cinderblockdeviceapi_for_test(test_case=self)
+
+    def _detach(self, instance_id, volume):
+        self.nova.volumes.delete_server_volume(instance_id, volume.id)
+        return wait_for_volume_state(
+            volume_manager=self.nova.volumes,
+            expected_volume=volume,
+            desired_state=u'available',
+            transient_states=(u'in-use', u'detaching'),
+        )
+
+    def _cleanup(self, instance_id, volume):
+        volume.get()
+        if volume.attachments:
+            self._detach(instance_id, volume)
+        self.cinder.volumes.delete(volume.id)
+
+    @require_virsh
+    def test_getdevicepath_error_without_udev(self):
+        """
+        ``getdevicepath`` on systems using the virtioblk driver raises
+        ``TimeoutException`` if ``/dev/disks/by-id/xxx`` does not appear within
+        5 seconds.
+        """
+        instance_id = self.blockdevice_api.compute_instance_id()
+        # Create volume
+        cinder_volume = self.cinder.volumes.create(
+            size=int(Byte(get_minimum_allocatable_size()).to_GiB().value)
+        )
+        self.addCleanup(self._cleanup, instance_id, cinder_volume)
+        volume = wait_for_volume_state(
+            volume_manager=self.cinder.volumes, expected_volume=cinder_volume,
+            desired_state=u'available', transient_states=(u'creating',))
+
+        # Suspend udevd before attaching the disk
+        [udev_process] = [
+            p for p in psutil.process_iter()
+            if p.name().endswith('-udevd')
+        ]
```

Comment:
:question: If you use a real for loop here, you could `break` when the process is found.